airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py

```diff
@@ -0,0 +1,240 @@
+            return None
+        elif self.concurrent_attempt < self.concurrent_max_retry:
+            # increment attempt
+            self.concurrent_attempt += 1
+            # try to execute previous request, it's handy because we can retry / each slice yielded
+            self.logger.warning(
+                f"The BULK concurrency limit has reached. Waiting {self.concurrent_interval_sec} sec before retry, atttempt: {self.concurrent_attempt}.",
+            )
+            sleep(self.concurrent_interval_sec)
+            # retry current `request`
+            return self.job_create(request)
+
+    def job_create(self, request: requests.Request) -> Optional[requests.Response]:
+        # create the server-side job
+        response = self.session.send(request)
+        # when the concurrent job takes place, we typically need to wait and retry, but no longer than 10 min.
+        if not self.has_running_concurrent_job(response):
+            # process created job response
+            return response if not self.job_check_for_errors(response) else None
+        else:
+            return self.job_retry_on_concurrency(request)
+
+    def log_job_status(self, bulk_job_id: str, status: str) -> None:
+        self.logger.info(f"The BULK Job: `{bulk_job_id}` is {status}.")
+
+    def job_check_status(self, url: str, bulk_job_id: str) -> Optional[str]:
+        # re-use of `self._session(*, **)` to make BULK Job status checks
+        request_args = {
+            "method": "POST",
+            "url": url,
+            "data": ShopifyBulkTemplates.status(bulk_job_id),
+            "headers": {"Content-Type": "application/graphql"},
+        }
+        response = self.session.request(**request_args)
+        # check for errors and status, return when COMPLETED.
+        if not self.job_check_for_errors(response):
+            status = response.json().get("data", {}).get("node", {}).get("status")
+            if status == ShopifyBulkStatus.COMPLETED.value:
+                self.log_job_status(bulk_job_id, status)
+                return response.json().get("data", {}).get("node", {}).get("url")
+            elif status == ShopifyBulkStatus.RUNNING.value:
+                self.log_job_status(bulk_job_id, status)
+                # wait for the `job_check_interval_sec` value in sec before check again.
+                sleep(self.job_check_interval_sec)
+                return self.job_check_status(url, bulk_job_id)
+            elif status == ShopifyBulkStatus.FAILED.value:
+                self.log_job_status(bulk_job_id, status)
+                raise ShopifyBulkExceptions.BulkJobFailed(
+                    f"The BULK Job: `{bulk_job_id}` exited with {status}, details: {self.job_check_for_errors(response)}",
+                )
+            elif status == ShopifyBulkStatus.TIMEOUT.value:
+                self.log_job_status(bulk_job_id, status)
+                raise ShopifyBulkExceptions.BulkJobTimout(
```

Comment:
should this be considered a config error or a system error? The error message makes me think this is a config error, but it's not obvious to me that'll always be the case

Reply:
This is exactly what it says: `ShopifyBulkExceptions.BulkJobTimout`. Neither the `config_error` nor `system_error`. This is the error on the  `Shopify` side. Ideally, we should retry with a smaller `slice_date_range`, but I've not faced such an error yet. We should fail here when we face such an error,  then handle it properly, and then convert it to the known `system/config` error.

Reply:
right. question is: is this actionable by the user? If so, it should be handled and turned into a config error. If not, the error message seems misleading to me

Reply:
Made it as a `config_errors` for all of the exceptions. It's actually manageable now, with the `GraphQL BULK Date Range in Days` spec property. Also, I've corrected the `Log Message` for the user, so it should be straightforward what to do.
Once we face some TIMEOUT errors, we'll be able to handle that with the `reduce the slice date range and retry` strategy.